File: C/Multicliente/servidor-multicliente.c
/*
 * Servidor de sockets TCP multicliente en C, usando pthreads.
 * Atiende múltiples clientes a la vez, un thread por conexión.
 * Equivalente a servidor-multicliente.py
 */



 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
 #include <arpa/inet.h>
 #include <pthread.h>

 #define PORT 65432
 #define BUFFER_SIZE 1024

typedef struct
{
    int client_fd;
    struct sockaddr_in address;
} datos_cliente;



void *atender_cliente(void *arg){
    datos_cliente *datos = (datos_cliente *)arg;
    int client_fd = datos -> client_fd;

    char ip_str[INET_ADDRSTRLEN];
    inet_ntop(AF_INET, &datos -> address.sin_addr,ip_str, INET_ADDRSTRLEN);
    int port = ntohs(datos -> address.sin_port);

    printf("[NUEVA CONEXION] %s:%d conectado.\n", ip_str, port);

    char buffer [BUFFER_SIZE];
    
    while (1)
    {
        memset(buffer, 0, BUFFER_SIZE);
        ssize_t bytes_leidos = read(client_fd, buffer, BUFFER_SIZE - 1);

        if(bytes_leidos < 0){
            printf("Eroor al leer del cliente");
            break;
        }

        if (bytes_leidos == 0)
        {
            printf("[DESCONECTADO] %s:%d cerro la conexion.\n", ip_str, port);
            break;
        }

        buffer[bytes_leidos] = '\0';
        printf("[%s:%d] dice: %s\n", ip_str, port, buffer);

        if (strncasecmp(buffer, "salir", 5) == 0 && bytes_leidos == 5)
        {
            const char *despedida = "Chau!";
            send(client_fd, despedida,strlen(despedida), 0);
            printf("[DESCONECTADO] %s:%d pidio salir.\n", ip_str, port);
            break;
        }
    
        char respuesta[BUFFER_SIZE];
        snprintf(respuesta, BUFFER_SIZE, "Servidor recibio: '%s'", buffer);
        send(client_fd, respuesta, strlen(respuesta), 0);
        
    }
    
    close(client_fd);
    free(datos);
    return NULL;
}

 int main(){

    int server_fd, client_fd;
    struct sockaddr_in address;
    socklen_t addr_len = sizeof(address);

    //---socket---

    server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0)
    {
        perror("Error al crear socket");
        exit(EXIT_FAILURE);
    }
    

    //---Reusar la direccion---

    int opt = 1;
    setsockopt(server_fd,SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    //---Configurar direccion de listen---

    memset(&address, 0 , sizeof(address));
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;

    address.sin_port = htons(PORT);


    //---Asociar socket a direccion/puerto---

    if(bind(server_fd, (struct sockaddr *)&address, sizeof(address))< 0) {
        perror("Error en bind");
        close(server_fd);
        exit(EXIT_FAILURE);
    }

    //---listen---

    if(listen(server_fd, 5) < 0){
        perror("Error en listen");
        close(server_fd);
        exit(EXIT_FAILURE);
    }

    printf("Servidor escuchando en el puerto %d... \n", PORT);

    while (1)
    {
        //---accept---
        client_fd = accept(server_fd, (struct sockaddr*)&address, &addr_len);
        if (client_fd < 0){
        perror("Error en accept");
        continue;
        }

        datos_cliente *datos = malloc(sizeof(datos_cliente));
        datos -> client_fd = client_fd;
        datos -> address = address;

        pthread_t hilo;

        if (pthread_create(&hilo, NULL, atender_cliente,datos) != 0)
        {
            perror("Error al crear el thread");
            free(datos);
            close(client_fd);
            continue;
        }

        pthread_detach(hilo);
        
    }


    close(server_fd);
    return 0;
 }

File: Python/Multicliente/servidor-multicliente.py


"""
Servidor de sockets TCP que atiende múltiples clientes a la vez
usando un thread por cada conexión, y acepta un comando especial
de administrador para apagarse remotamente.
 
Al apagarse, avisa a todos los clientes conectados antes de cerrar
el proceso (los threads son daemon, así que mueren apenas main()
termina, pero primero les mandamos un mensaje de aviso).
"""

import threading
import socket
import os

HOST = "127.0.0.1"  
PORT = 65432

ADMIN_PASSWORD = os.environ.get("ADMIN_PASSWORD", "admin123")

shutdown_event = threading.Event()

clientes_activos = []
clientes_lock = threading.Lock()



def atender_cliente(conn, addr, server_socket):
    
    """
    Corre en un thread por cada cliente. El PRIMER mensaje siempre se
    interpreta como intento de autenticación: "ADMIN:<contraseña>" para
    loguearse como admin, o cualquier otra cosa para ser un cliente normal.
 
    Un admin autenticado puede después mandar mensajes de chat como
    cualquier cliente, O el comando "shutdown" en cualquier momento
    para apagar el servidor.
    """

    print(f"[NUEVA CONEXION] {addr} conectado")
    es_admin = False
    
    with clientes_lock:
        clientes_activos.append((conn,addr))
        
    try:
    
        with conn:
            #--- Primer mensaje : intento de autenticacion ---
            
            data = conn.recv(1024)
            if not data:
                print (f"[DESCONECTADO] {addr} cerro la conexion antes de identificarse.")
                return
            
            primer_mensaje = data.decode("utf-8").strip()
            partes = primer_mensaje.split(":")
            
            if len(partes) == 2 and partes[0] == "ADMIN":
                #print(f"DEBUG: recibido={repr(partes[1])} | esperado={repr(ADMIN_PASSWORD)}")
                if partes[1] == ADMIN_PASSWORD:
                    es_admin = True
                    print(f"[ADMIN] {addr} se autentico correctamente.")
                    conn.sendall("Ok, autenticado como admin. podes chatear o mandar shutdown".encode("utf-8"))
                
                else:
                    print(f"[ADMIN] {addr} intento autenticarse con contraseña incorrecta.")
                    conn.sendall("Contraseña incorrecta.".encode("utf-8"))
                    return
            
            else:
                print(f"[{addr}] dice: {primer_mensaje}")
                if primer_mensaje.lower()== "salir":
                    conn.sendall("Chau!".encode("utf-8"))
                    print(f"[DESCONECTADO] {addr} pidio salir.")
                    return
                
                respuesta = f"Servidor recibio: '{primer_mensaje}'"
                conn.sendall(respuesta.encode("UTF-8"))
        
            #--- Loop normal -----
            
            while True:
                    data = conn.recv(1024)
                    
                    if  not data:
                        print(f"[DESCONECTADO] {addr} cerró la conexion")
                        break
                    
                    mensaje = data.decode('utf-8').strip()
                    
                    if es_admin and mensaje.lower() == "shutdown":
                        print(f"[ADMIN] {addr} pidio apagar el servidor")
                        conn.sendall("OK, apagando servidor...".encode("UTF-8"))
                        shutdown_event.set()
                        break
                    
                    print(f"[{addr}] dice: {mensaje}")
                    
                    if mensaje.lower() == "salir":
                        conn.sendall("Chau!".encode("utf-8"))
                        print(f"[DESCONECTADO] {addr} pidio salir.")
                        break
                    
                    respuesta = f"Servidor recibio: '{mensaje}'"
                    conn.sendall(respuesta.encode("utf-8"))
    finally:
        
        with clientes_lock:
            if(conn,addr) in clientes_activos:
                clientes_activos.remove((conn,addr))
                

def avisar_y_cerrar():
    
    with clientes_lock:
        copia = list(clientes_activos)
        
    for conn,add in copia:
        try:
            conn.sendall("El servidor se esta apagando. Chau!".encode('utf-8'))
        except OSError:
            pass
    

def main():     

    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_socket.bind((HOST, PORT))
    server_socket.listen()
    
    server_socket.settimeout(1.0)
    print(f"Servidor escuchando en {HOST}:{PORT}...")

    try:
         
        while not shutdown_event.is_set():
            try:
                conn, addr = server_socket.accept()
                
            except socket.timeout:
                continue
            
            except OSError:
                break
            
            hilo = threading.Thread(target= atender_cliente, args=(conn,addr,server_socket),daemon=True)
            hilo.start()
            
            print(f"[CLIENTES ACTIVOS] {threading.active_count() - 1}")
    
    except KeyboardInterrupt:
        print("\n[APAGANDO] Ctrl+c detectado, cerrando servidor...")
        shutdown_event.set()
        
    print("[APAGANDO] Avisando a clientes conectados...")
    avisar_y_cerrar()
    
    
        
    server_socket.close()
    print("[APAGADO] Servidor apagado.")
    print("Servidor finalizado.")
    
                
if __name__ == "__main__":
    main() 